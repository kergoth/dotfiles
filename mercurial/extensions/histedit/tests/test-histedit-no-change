#!/bin/sh

# test for issue #6:
# editing a changeset without any actual change would corrupt the repository

. "$TESTDIR/histedit-helpers.sh"

cat >> $HGRCPATH <<EOF
[extensions]
hgext.graphlog=
hg_histedit=$(echo $(dirname $(dirname $0)))/hg_histedit.py
EOF

initrepo ()
{
    dir="$1"
    comment="$2"

    if [ -n "${comment}" ]; then
        echo % ${comment}
        echo % ${comment} | sed 's:.:-:g'
    fi

    hg init ${dir}
    cd ${dir}
    for x in a b c d e f ; do
        echo $x > $x
        hg add $x
        hg ci -m $x
    done
}

geneditor ()
{
    # generate an editor script for selecting changesets to be edited

    choice=$1  # changesets that should be edited (using sed line ranges)

    cat <<EOF | sed 's:^....::'
    #!/bin/sh

    # editing the rules, replacing 'pick' with 'edit' for the chosen lines
    sed '${choice}s:^pick:edit:' \$1 > \${1}.tmp
    mv \${1}.tmp \$1

    # displaying the resulting rules, minus comments and empty lines
    sed '/^#/d;/^$/d;s:^:| :' \$1 >&2
EOF
}

startediting ()
{
    # begin an editing session

    choice="$1"  # changesets that should be edited
    number="$2"  # number of changesets considered (from tip)
    comment="$3"

    geneditor "${choice}" > edit.sh
    chmod +x edit.sh

    echo % start editing the history ${comment}
    HGEDITOR=./edit.sh hg histedit -- -${number} 2>&1 | fixbundle
}

continueediting ()
{
    # continue an edit already in progress

    editor="$1"  # message editor when finalizing editing
    comment="$2"

    echo % finalize changeset editing ${comment}
    HGEDITOR=${editor} hg histedit --continue 2>&1 | fixbundle
}

graphlog ()
{
    comment="${1:-log}"

    echo % "${comment}"
    hg glog --template '{rev} {node} \"{desc|firstline}\"\n'
}



initrepo r1 "test editing with no change"
graphlog "log before editing"
startediting 2 3 "(not changing anything)" # edit the 2nd of 3 changesets
continueediting true "(leaving commit message unaltered)"

echo "% check state of working copy"
hg id

graphlog "log after history editing"


cd ..
initrepo r2 "test editing with no change, then abort"
graphlog "log before editing"
startediting 1,2 3 "(not changing anything)" # edit the 1st two of 3 changesets
continueediting true "(leaving commit message unaltered)"
graphlog "log after first edit"

echo % abort editing session
hg histedit --abort 2>&1 | fixbundle

graphlog "log after abort"

echo % EOF
