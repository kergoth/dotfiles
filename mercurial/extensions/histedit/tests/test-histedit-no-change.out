% test editing with no change
-----------------------------
% log before editing
@  5 652413bf663ef2a641cab26574e46d5f5a64a55a "f"
|
o  4 e860deea161a2f77de56603b340ebbb4536308ae "e"
|
o  3 055a42cdd88768532f9cf79daa407fc8d138de9b "d"
|
o  2 177f92b773850b59254aa5e923436f921b55483b "c"
|
o  1 d2ae7f538514cd87c17547b0de4cea71fe1af9fb "b"
|
o  0 cb9a9f314b8b07ba71012fcdbc544b5a4d82ff5b "a"

% start editing the history (not changing anything)
| pick 055a42cdd887 3 d
| edit e860deea161a 4 e
| pick 652413bf663e 5 f
0 files updated, 0 files merged, 2 files removed, 0 files unresolved
abort: Make changes as needed, you may commit or record as needed now.
When you are finished, run hg histedit --continue to resume.
% finalize changeset editing (leaving commit message unaltered)
1 files updated, 0 files merged, 0 files removed, 0 files unresolved
% check state of working copy
652413bf663e tip
% log after history editing
@  5 652413bf663ef2a641cab26574e46d5f5a64a55a "f"
|
o  4 e860deea161a2f77de56603b340ebbb4536308ae "e"
|
o  3 055a42cdd88768532f9cf79daa407fc8d138de9b "d"
|
o  2 177f92b773850b59254aa5e923436f921b55483b "c"
|
o  1 d2ae7f538514cd87c17547b0de4cea71fe1af9fb "b"
|
o  0 cb9a9f314b8b07ba71012fcdbc544b5a4d82ff5b "a"

% test editing with no change, then abort
-----------------------------------------
% log before editing
@  5 652413bf663ef2a641cab26574e46d5f5a64a55a "f"
|
o  4 e860deea161a2f77de56603b340ebbb4536308ae "e"
|
o  3 055a42cdd88768532f9cf79daa407fc8d138de9b "d"
|
o  2 177f92b773850b59254aa5e923436f921b55483b "c"
|
o  1 d2ae7f538514cd87c17547b0de4cea71fe1af9fb "b"
|
o  0 cb9a9f314b8b07ba71012fcdbc544b5a4d82ff5b "a"

% start editing the history (not changing anything)
| edit 055a42cdd887 3 d
| edit e860deea161a 4 e
| pick 652413bf663e 5 f
0 files updated, 0 files merged, 3 files removed, 0 files unresolved
abort: Make changes as needed, you may commit or record as needed now.
When you are finished, run hg histedit --continue to resume.
% finalize changeset editing (leaving commit message unaltered)
0 files updated, 0 files merged, 0 files removed, 0 files unresolved
abort: Make changes as needed, you may commit or record as needed now.
When you are finished, run hg histedit --continue to resume.
% log after first edit
o  5 652413bf663ef2a641cab26574e46d5f5a64a55a "f"
|
o  4 e860deea161a2f77de56603b340ebbb4536308ae "e"
|
@  3 055a42cdd88768532f9cf79daa407fc8d138de9b "d"
|
o  2 177f92b773850b59254aa5e923436f921b55483b "c"
|
o  1 d2ae7f538514cd87c17547b0de4cea71fe1af9fb "b"
|
o  0 cb9a9f314b8b07ba71012fcdbc544b5a4d82ff5b "a"

% abort editing session
2 files updated, 0 files merged, 0 files removed, 0 files unresolved
% log after abort
@  5 652413bf663ef2a641cab26574e46d5f5a64a55a "f"
|
o  4 e860deea161a2f77de56603b340ebbb4536308ae "e"
|
o  3 055a42cdd88768532f9cf79daa407fc8d138de9b "d"
|
o  2 177f92b773850b59254aa5e923436f921b55483b "c"
|
o  1 d2ae7f538514cd87c17547b0de4cea71fe1af9fb "b"
|
o  0 cb9a9f314b8b07ba71012fcdbc544b5a4d82ff5b "a"

% EOF
